xapian-core/common/serialise-double.cc

```diff
@@ -38,175 +40,51 @@
 
 using namespace std;
 
-// The serialisation we use for doubles is inspired by a comp.lang.c post
-// by Jens Moeller:
-//
-// https://groups.google.com/group/comp.lang.c/browse_thread/thread/6558d4653f6dea8b/75a529ec03148c98
-//
-// The clever part is that the mantissa is encoded as a base-256 number which
-// means there's no rounding error provided both ends have FLT_RADIX as some
-// power of two.
-//
-// FLT_RADIX == 2 seems to be ubiquitous on modern UNIX platforms, while
-// some older platforms used FLT_RADIX == 16 (IBM machines for example).
-// FLT_RADIX == 10 seems to be very rare (the only instance Google finds
-// is for a cross-compiler to some TI calculators).
-
-#if FLT_RADIX == 2
-# define MAX_MANTISSA_BYTES ((DBL_MANT_DIG + 7 + 7) / 8)
-# define MAX_EXP ((DBL_MAX_EXP + 1) / 8)
-# define MAX_MANTISSA (1 << (DBL_MAX_EXP & 7))
-#elif FLT_RADIX == 16
-# define MAX_MANTISSA_BYTES ((DBL_MANT_DIG + 1 + 1) / 2)
-# define MAX_EXP ((DBL_MAX_EXP + 1) / 2)
-# define MAX_MANTISSA (1 << ((DBL_MAX_EXP & 1) * 4))
-#else
-# error FLT_RADIX is a value not currently handled (not 2 or 16)
-// # define MAX_MANTISSA_BYTES (sizeof(double) + 1)
-#endif
-
-static int base256ify_double(double &v) {
-    int exp;
-    v = frexp(v, &exp);
-    // v is now in the range [0.5, 1.0)
-    --exp;
-#if FLT_RADIX == 2
-    v = scalbn(v, (exp & 7) + 1);
-#else
-    v = ldexp(v, (exp & 7) + 1);
-#endif
-    // v is now in the range [1.0, 256.0)
-    exp >>= 3;
-    return exp;
-}
-
-std::string serialise_double(double v)
+string serialise_double(double v)
 {
-    /* First byte:
-     *  bit 7 Negative flag
-     *  bit 4..6 Mantissa length - 1
-     *  bit 0..3 --- 0-13 -> Exponent + 7
-     *            \- 14 -> Exponent given by next byte
-     *             - 15 -> Exponent given by next 2 bytes
-     *
-     * Then optional medium (1 byte) or large exponent (2 bytes, lsb first)
-     *
-     * Then mantissa (0 iff value is 0)
-     */
-
-    bool negative = (v < 0.0);
-
-    if (negative) v = -v;
-
-    int exp = base256ify_double(v);
-
-    string result;
-
-    if (exp <= 6 && exp >= -7) {
-	unsigned char b = static_cast<unsigned char>(exp + 7);
-	if (negative) b |= static_cast<unsigned char>(0x80);
-	result += char(b);
-    } else {
-	if (exp >= -128 && exp < 127) {
-	    result += negative ? char(0x8e) : char(0x0e);
-	    result += char(exp + 128);
-	} else {
-	    if (exp < -32768 || exp > 32767) {
+	int exp;
```

Comment:
You indentation doesn't match Xapian conventions (the `xapian-check-patch` script doesn't currently detect this case unfortunately) - probably you have your editor set wrongly.  In most cases, each indentation level should be 4 columns, you should have a tab character set to move to the next multiple of 8 columns, and indents should be "tab-filled" (so a single indent is 4 spaces; double indent a tab, and in general the indent is zero or more tabs then 0 or 4 spaces).

Reply:
I'm still not quite sure about the indentation, but I think the latest commit does follow what you say.

Reply:
There are still some problems, but I'll flag them individually.